### DISCUSSION TITLE
Speed of person / object detection

### DISCUSSION THREAD

**Discussion by LaurenceGough** [blakeblackshear/frigate#3531] [Ask A Question]
Hi all,

I was hoping the Google Coral USB accelerator would speed up object detection. Inference speed has dropped from 400-600ms to down to 10-20ms, which is great. However I was disappointed to see the same issue remains when checking the debug feed. Despite people being in view walking by (alley way with a brief view or at night time walking by) and that they were being processed for object recognition via the green image detection box was around them for at least half a second there was no object detected still.

Is there a minimum time for an object (person) to be present before marking as detected set anywhere? I couldn't see it when checking the config options.

I have tried playing with min_score which was the only thing I could find related, unless I'm being blind.

Many thanks in advance

Edited to correct my mistake of saying boundary box when I meant object searching green box. Cheers Nick.

**Comment by NickM-27** [comment]
The green box is not a bounding box (or I guess it kind of is, but it isn't for any particular object), it is the region which is sent for object detection. Odds are you have some filter which is blocking out this person from being considered a true detection

**Reply by LaurenceGough** [reply]
> The green box is not a bounding box (or I guess it kind of is, but it isn't for any particular object), it is the region which is sent for object detection. Odds are you have some filter which is blocking out this person from being detected


Doh, thanks I'm using the wrong terminology again, I meant the searching box. If a person mask is in the green search area but the person wasn't in that mask that should still work right?

I will remove the masks and try again, they are as small as possible and pretty tiny, they were just because it kept picking up small posts as people.

**Reply by NickM-27** [reply]
The masks filter based on the bottom-center of the persons bounding box. You may also want to look into the min_area and max_area filters as that could be blocking it. 

If you post your config it may be easiest, it's difficult to say for sure what's going on without seeing the results of an inference and what it sees, but if it's running detection (green region) and not returning anything than the person is either not being detected or filtered out, there will be no delay there due to latency or anything.

**Comment by blakeblackshear** [comment]
> Is there a minimum time for an object (person) to be present before marking as detected set anywhere?

It takes at least 3 frames where the median is above the threshold to create an event, but you should see bounding boxes for detections when watching debug in real time even if the threshold isn't met. If not, it's not detecting an object or it's being filtered out.

**Comment by LaurenceGough** [comment]
Thanks all. That's great to know. I've removed all person masks apart from one in some plants where no one can get to and set min_area to 0 and max_area to a massive number  (both were unset before). But according to the config guide this was the default anyway.  The person masks I don't feel were in the way but to be safe I removed them.

I've set min_score to 0.05 now from min_score 0.1. 
The description is "minimum score for the object to initiate tracking" which sounds just like what I'm after. Heck, maybe I should set that to 0? 

I mean 3 frames at 5fps is 0.3ms right? So if they are there for 0.3 of a second it should detect them, if all other conditions are met. 

I'll see how it goes now.

Now I have a Coral I also tried a different tensor library fine tuned for CCTV images but seemed the same sadly.

**Reply by NickM-27** [reply]
> I mean 3 frames at 5fps is 0.3ms right? So if they are there for 0.3 of a second it should detect them, if all other conditions are met.

Not quite. At 5 frames per second that is 1 frame every 0.2 seconds. So 0.6 seconds (in ideal conditions) to get a detection.

> I've set min_score to 0.05 now from min_score 0.1.

That seems extremely low TBH, but if it doesn't work then that will be a clue that it is barking up the wrong tree, I'd definitely recommend increasing that after this is figured out. Unless the people are very obscured then I don't see why it wouldn't work at normal settings.

**Reply by LaurenceGough** [reply]
Doh, thanks... my maths is so bad hahaha. Ah 0.6 of a second or 600ms is a bit longer than I thought perhaps this is it as a second is always longer than you think, and as you say that's ideal conditions for that too.

I wonder if this could be exposed as an option in the future? I've been having some trouble recently so it's key to know when someone has been there rather than having to review hours of footage.. I mean it's pretty much the whole reason I got the Coral and Frigate setup.

The camera is only 1080p but sometimes the people can be quite obscured by other things such as vehicles (I am not tracking vehicles).

I wonder what the harm is in having the min_score too low, apart from the Coral being used a bit more? I see more "post tracking" as a person (Frigate loves detecting posts as people ;)) with my other camera which is just using the default min_score setting.

Another thought I just had is increasing the fps, I can probably get away with that now I am using the Coral. I'll up the FPS slightly now. To be fair this is probably the best solution now I have this info about how long it takes.

Many thanks

**Reply by NickM-27** [reply]
> I wonder if this could be exposed as an option in the future? I've been having some trouble recently so it's key to know when someone has been there rather than having to review hours of footage.. I mean it's pretty much the whole reason I got the Coral and Frigate setup.

I don't think that would be an option, I imagine that would have way too many false positives. If you want quick detection then increasing your `detect -> fps` is the way to go (which will use the coral more, obviously)

> I wonder what the harm is in having the min_score too low, apart from the Coral being used a bit more? I see more "post tracking" as a person (Frigate loves detecting posts as people ;)) with my other camera which is just using the default min_score setting.

Right, I think 0.3 makes sense as the lowest option, but really do what works best for your case

**Comment by LaurenceGough** [comment]
So after reviewing the debug screen a bit more today, the object box for person can even go orange from blue which I believe means that the detection confidence was above the threshold set but no event is recorded. I feel like it has a minimum time before an event is recorded for a person being detected. My notifications also don't work which rely on an mqtt event.

I've seen this behaviour before. 

Any ideas on how to correct this so that as soon as the object detection boundary box goes orange it records and fires mqtt events?

Thank you :).

**Reply by NickM-27** [reply]
blue means the object was seen but it not currently seen. Orange means it is being tracked. In my usage MQTT & event fires immediately after the 3 frames. 

To be able to help I think we'd either need a screen recording of debugging screen or a an event mp4 recording (can send over email or Gdrive) along with the config to be able to diagnose if something off is happening

**Comment by 1990marco1990** [comment]
I have the same problem with a delay of 3-5 seconds to detect a person.I would use my camera for lights on if person detect but its very bad. Can you help me?

**Reply by LaurenceGough** [reply]
Increasing the resolution of the detect stream and increasing to 8fps helped a lot. I honestly think the res made the most difference helps accuracy it seems a lot on both of my cameras. Sadly as hw decoding is broken currently on Pis it uses a lot more CPU too.

**Comment by 1990marco1990** [comment]
Ok I have test it. Its a little bit better but not enough for lights on and off.

**Reply by NickM-27** [reply]
I'd recommend making your own issue, I've had great success with lights on off. 

If you don't have a coral then that adds a lot of delay as well.
